API: Have MultiIndex consturctors always return a MI

TomAugspurger opened the pull request:
This removes the special case for MultiIndex constructors returning
an Index if all the levels are length-1. Now this will return a
MultiIndex with a single level.

This is a backwards incompatabile change, with no clear method for
deprecation, so we're making a clean break.

Old Behavior:

```python
In [1]: import pandas as pd

In [2]: pd.MultiIndex.from_tuples([('a',), ('b',)])
Out[2]: Index(['a', 'b'], dtype='object')

```

New:

```python
In [1]: import pandas as pd
In [2]: pd.MultiIndex.from_tuples([('a',), ('b',)])
Out[2]:
MultiIndex(levels=[['a', 'b']],
           labels=[[0, 1]])
```

Closes #17178

jreback commented:
I don't think this is a good idea at all. Why are you not simply converting these 1-tuples? This is just adding complexity.

TomAugspurger commented:
> I don't think this is a good idea at all. Why are you not simply converting these 1-tuples?

For the categorical case, we need to treat `categories=[1, 2, 3]` differently than `categories=[(1,), (2,), (3,)]`. I could add additional code there to special case categories of length 1, but the fix here is much cleaner.

> This is just adding complexity.

Perhaps a bit? I'm not sure an additional keyword + a one line change to the MI constructor is a maintenance burden.

We've been trying to remove all the areas where the return type is dependent on the values, like it is here. That said, I don't think this is worth breaking API over, since it's been this way for so long. It's the mirror behavior of `Index` with tuples:

```python
In [9]: pd.Index([(1, 2), (3, 4)])
Out[9]:
MultiIndex(levels=[[1, 3], [2, 4]],
           labels=[[0, 1], [0, 1]])

In [10]: pd.Index([(1, 2), (3, 4)], tupleize_cols=False)
Out[10]: Index([(1, 2), (3, 4)], dtype='object')
```

The constructors try to be smart about what you want, but give you control if you know better what you want.

jbrockmendel commented:
The ability to have a 1-level MI would be nice.  It's a step towards making MI and non-MI quack alike.

jreback commented:
the ``tupleize_cols`` kw to ``Index`` is not a good example. that should be removed as well. 
This is just not a good solution here. If its a problem in ``Categorical``, then more introspection is warranted.

Adding a keyword IS adding more complexity in the constructor here.

TomAugspurger commented:
> the tupleize_cols kw to Index is not a good example. that should be removed as well.

I disagree, but opened #17246 for the overall issue.

> If its a problem in Categorical, then more introspection is warranted.

Categorical is an *instance* of this problem surfacing. I just really don't think it should be on the user to ensure that their tuples are all length > 1 every time they want to use a MultiIndex. e.g. fixing `hash_tuples` to work on tuples of length 1. I'm not sure what you mean by "more introspection is warranted."

jreback commented:
> the tupleize_cols kw to Index is not a good example. that should be removed as well.

kwargs just adding technical debt here. I am -1 on this soln.

jreback commented:
@TomAugspurger the problem here is you are trying to use a well-defined public API internally. Instead you merely need to use (and possibly modify) ``ensure_index`` which is the right place for this.

TomAugspurger commented:
I want to change the public API; I don't think the return type should depend on the data. But I don't want to break existing code, hence the keyword.

jreback commented:
my point is why don't you just change it. I don't think

```
In [1]: import pandas as pd
In [2]: pd.MultiIndex.from_tuples([('a',), ('b',)])
Out[2]: Index(['a', 'b'], dtype='object')
```

make any senses anyhow.

TomAugspurger committed:
API: Have MultiIndex constructors return MI

This removes the special case for MultiIndex constructors returning
an Index if all the levels are length-1. Now this will return a
MultiIndex with a single level.

This is a backwards incompatabile change, with no clear method for
deprecation, so we're making a clean break.

Closes #17178

jorisvandenbossche commented:
@jreback just to be sure I understand you correctly: you are opposed to adding a new kwarg (which indeed adds complexity), but you are OK with actually changing the behaviour so that `MultiIndex.from_tuples` with 1-length tuples returns a MultiIndex instead of an Index ?

Because in https://github.com/pandas-dev/pandas/issues/17178#issuecomment-320459446 you said that the current implementation (return plain Index from `MultiIndex.from_tuples` given 1-length tuples) "is correct" (and it is indeed also tested explicitly, so at the time this behaviour was added on purpose)

jreback commented:
I think its clear that adding a keyword to simply avoid an existing special case is very odd. better to simply fix ``MultiIndex.from_tuples`` to simply always return a ``MultiIndex``, which eliminates the current special case.

jorisvandenbossche commented:
OK, thanks for the clarification! Then I agree with just fixing the special case (although this is a back incompatible change)

pep8speaks commented:
Hello @TomAugspurger! Thanks for updating the PR.

 - In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/be177912d46bcb985216a7aeda3edd2b1e4f7724/pandas/core/frame.py), following are the PEP8 issues :

> [Line 1160:80](https://github.com/pandas-dev/pandas/blob/be177912d46bcb985216a7aeda3edd2b1e4f7724/pandas/core/frame.py#L1160): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

TomAugspurger commented:
Updated, if anyone has a chance to look. We made use of this in quite a few places, and all the extra if checks are a bit annoying, but I think worth it in the end to know that `MultiIndex` will always return a MultiIndex.

jorisvandenbossche commented:
If we find the if/else statements too annoying, we could also have a private 'squeeze' method on a MultiIndex that returns a plain Index if it has only one level and otherwise leaves it as a MultiIndex. 
But it seems it are not that many places that you needed to change it.

TomAugspurger commented:
Yeah, I think not enough places to warrant a private method.

On Wed, Aug 23, 2017 at 3:22 PM, Joris Van den Bossche <
notifications@github.com> wrote:

> If we find the if/else statements too annoying, we could also have a
> private 'squeeze' method on a MultiIndex that returns a plain Index if it
> has only one level and otherwise leaves it as a MultiIndex.
> But it seems it are not that many places that you needed to change it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/17236#issuecomment-324451265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIslQkD1lPIBkC5KWyzTGMp-wMUTwks5sbInrgaJpZM4O1Yv7>
> .
>

jreback commented on pandas/core/frame.py:
```diff
@@ -1156,8 +1156,12 @@ def from_records(cls, data, index=None, exclude=None, columns=None,
 
-                    result_index = MultiIndex.from_arrays(
-                        [arrays[i] for i in to_remove], names=index)
+                    if len(to_remove) == 1:
```
I think you should be using ``_ensure_index`` here.

jreback commented on pandas/core/frame.py:
```diff
@@ -3000,7 +3004,10 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
                     to_remove.append(col)
             arrays.append(level)
 
-        index = MultiIndex.from_arrays(arrays, names=names)
```
same (or at least that's the intent); otherwise maybe should have an option / another helper function to avoid this repetition of code.

jreback commented on pandas/io/parsers.py:
```diff
@@ -1444,7 +1444,14 @@ def _agg_index(self, index, try_parse_dates=True):
             arrays.append(arr)
 
-        index = MultiIndex.from_arrays(arrays, names=self.index_names)
+        if len(arrays) > 1:
```
yeah this too

jreback commented on pandas/io/parsers.py:
```diff
@@ -1808,7 +1815,10 @@ def read(self, nrows=None):
                 arrays.append(values)
 
-            index = MultiIndex.from_arrays(arrays)
+            if self._reader.leading_cols == 1:
```
same

jreback commented on pandas/io/parsers.py:
```diff
@@ -3138,9 +3148,13 @@ def _get_empty_meta(columns, index_col, index_names, dtype=None):
-        index = [Series([], dtype=dtype[index_name])
-                 for index_name in index_names]
-        index = MultiIndex.from_arrays(index, names=index_names)
+        if len(index_names) == 1:
```
same

jreback reviewed (changes requested):
left some comments. looks generally, but I think can easily wrap then if/else code (better to use _ensure_index, which may need some modification)

TomAugspurger commented:
Updated. Instead of using `_ensure_index` I added a small helper method `_index_from_sequences` that does the `if len(x) == 1` checking. `_ensure_index` seemed to be sufficiently different from what I needed to warrant a new method.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4007,6 +4007,33 @@ def invalid_op(self, other=None):
 Index._add_comparison_methods()
 
 
+def _index_from_sequences(sequences, names=None):
```
can you rename to ``_ensure_index_from_sequences`` for consistency.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4007,6 +4007,33 @@ def invalid_op(self, other=None):
+
+
 def _ensure_index(index_like, copy=False):
     if isinstance(index_like, Index):
```
can you add a doc-string here as well

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -446,7 +449,10 @@ def _slow_pivot(index, columns, values):
 
 def unstack(obj, level, fill_value=None):
     if isinstance(level, (tuple, list)):
-        return _unstack_multiple(obj, level)
```
a comment on why this if is needed here

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -311,10 +311,13 @@ def _unstack_multiple(data, clocs):
-                             labels=rlabels + [comp_ids],
-                             names=rnames + ['__placeholder__'],
-                             verify_integrity=False)
+    if rlocs == []:
```
a comment on why the if is needed

jreback commented on pandas/core/strings.py:
```diff
@@ -1452,7 +1452,10 @@ def cons_row(x):
                 result = list(result)
-                return MultiIndex.from_tuples(result, names=name)
+                out = MultiIndex.from_tuples(result, names=name)
+                if out.nlevels == 1:
```
a comment on why this is needed

jreback commented on pandas/io/parsers.py:
```diff
@@ -1444,7 +1445,16 @@ def _agg_index(self, index, try_parse_dates=True):
 
-        index = MultiIndex.from_arrays(arrays, names=self.index_names)
+        names = self.index_names
+        index = _index_from_sequences(arrays, names)
```
this seems not needed here

jreback replied:
comment about what you are interpreting here

jreback commented on pandas/util/testing.py:
```diff
@@ -1910,6 +1910,9 @@ def keyfunc(x):
     # convert tuples to index
```
can you add a comment here

jreback reviewed (changes requested):
couple of comments

TomAugspurger committed:
fixup! API: Have MultiIndex constructors return MI

TomAugspurger commented:
I think all the comments have been addressed.

jreback reviewed (approved):
lgtm

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -273,6 +273,27 @@ named ``.isna()`` and ``.notna()``, these are included for classes ``Categorical
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previous versions would automatically squeeze a ``MultiIndex`` with length-one
+``levels`` down to an ``Index``:
```
can you add the issue number here (the PR or issue if we have one)

jreback replied:
this one as well at the top (and not the bottom)

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4007,7 +4007,68 @@ def invalid_op(self, other=None):
+    MultiIndex(levels=[['a'], ['a', 'b']],
+               labels=[[0, 0], [0, 1]],
+               names=['L1', 'L2'])
+    """
```
can you add a See Also to _ensure_index (and vice-versa).

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -446,7 +450,12 @@ def _slow_pivot(index, columns, values):
     if isinstance(level, (tuple, list)):
-        return _unstack_multiple(obj, level)
+        if len(level) == 1:
+            # unstack_multiple only handles MultiIndexes,
```
maybe reverse this condition to make it more clear

```
if len(level) != 1:
    return .....
level = level[0]
```

jreback reviewed (changes requested):
small comments. merge when pushed / green.

TomAugspurger committed:
Update for comments

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -1,6 +1,9 @@
-                                 InvalidIndexError)
+from pandas.core.indexes.base import (Index,
+                                      _new_Index,
+                                      _ensure_index,
```
this shouldn't need a noqa (if its line length, break it on the prarens

TomAugspurger replied:
I think some of these are unused  and just there to export as part of the API.

jreback replied:
k

jreback reviewed (approved):
tiny comments
